### PR TITLE
FM2-431-WIP: Add support to search the Encounter resource by EncounterType

### PR DESCRIPTION
**Pull request by jnsereko**
## Description of what I changed
Add support to search the Encounter resource by EncounterType

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-431

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/EncounterFhirResourceProvider.java [hunk 1]
```diff
@@ -135,6 +136,7 @@ public IBundleProvider searchEncounter(@OptionalParam(name = Encounter.SP_DATE)
 	        @OptionalParam(name = Encounter.SP_PATIENT, chainWhitelist = { "", Patient.SP_IDENTIFIER, Patient.SP_GIVEN,
 	                Patient.SP_FAMILY, Patient.SP_NAME }, targetTypes = Patient.class) ReferenceAndListParam patientParam,
 	        @OptionalParam(name = Encounter.SP_RES_ID) TokenAndListParam id,
+	        @OptionalParam(name = Encounter.SP_TYPE) CodeableConcept type,
```
Comment (jnsereko) [inline api/src/main/java/org/openmrs/module/fhir2/providers/r3/EncounterFhirResourceProvider.java:139]: Hope this is the right track. 
Should the type be `org.hl7.fhir.r4.model.CodeableConcept` or `ReferenceAndListParam`
cc @ibacher @mozzy11 @pmanko

**Comment by jnsereko** [comment]
Hey @ibacher @mozzy11 @pmanko.
I am running into an error [Unsupported type parameter: org.hl7.fhir.r4.model.CodeableConcept](https://pastebin.com/4Z9MgGqX) which has made me question whether am still on the right track,  
If this is the correct algorithm. How can I solve this error. Is there any mapping I need to create, like interaction with the database? My aim is to make all existing tests first pass before adding tests for this feature.

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/EncounterFhirResourceProvider.java (hunk 1 above)
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r3/EncounterFhirResourceProvider.java:139]: It should actually be `TokenAndListParam` like `id` above. FHIR's search API has [different types](https://www.hl7.org/fhir/search.html#ptypes) than FHIR's resource representations, but in general if the datatype is a `CodeableConcept` the search parameter will be a `token`, which is the type given for [the `type` search parameter](https://www.hl7.org/fhir/encounter.html#search)

**Comment by jnsereko** [comment]
Closing this because its a duplicate.  Branch name was wrongly created. on updating commit, instead, an new PR was created https://github.com/openmrs/openmrs-module-fhir2/pull/371

**Pull request closed without merging**